### PR TITLE
feat!: embed the transaction which spent the input DBC within the Dbc struct

### PR DESCRIPTION
**Pull request by bochaco**

### CODE REVIEW

File: src/dbc.rs [hunk 1]
```diff
@@ -345,8 +361,9 @@ pub(crate) mod tests {
             .build(rng)?;
 
         for (key_image, tx) in dbc_builder.inputs() {
-            dbc_builder =
-                dbc_builder.add_spent_proof_share(spentbook_node.log_spent(key_image, tx)?);
+            dbc_builder = dbc_builder
+                .add_spent_proof_share(spentbook_node.log_spent(key_image, tx.clone())?)
+                .add_spent_transaction(tx);
```
Comment (maqi) [inline src/dbc.rs:366]: this looks like there will be `one transaction per proof_share` get added, which means X copies of duplicated tx get pushed to the vector?

File: src/builder.rs [hunk 2]
```diff
@@ -401,6 +407,15 @@ impl DbcBuilder {
         // note that we do this just once for entire Tx, not once per output Dbc.
         TransactionVerifier::verify(verifier, &self.transaction, &spent_proofs)?;
 
+        // verify there is a maching spent transaction for each spent_proof
```
[the remark this answers is not in the document]
Comment (bochaco) [reply src/builder.rs:410]: I don't think it's necessary, what it's important is that for each proof there is a TX to verify the hash against, if there are more TXs shouldn't matter I believe?

File: src/dbc.rs [hunk 3]
```diff
@@ -322,7 +338,7 @@ pub(crate) mod tests {
...
...dda7c87e1d3b30254c25aba701ff6eeb9ebe8800000000000000011ca2ae5ab0ed3dc52cc9cd341c4a482e9d9f5f2d81d981a11032821603b863baa19a900087052a2f799092f29d7e2c8b548303a73305fc9c94eada44863144d84b6f9088a6280ff1b54cf540dc6c4ded457121351f13df5ffd94cb9764079dede3220daca7d85f61d5092eaca480b7b500000000000000013f80a95efa16af78eceb4d770980e57c9f2ef4939db4fd6233e71179b281c30c000000000000000117b4bfcd5437771b00ce5d6fc6f604576e621cd12754539817b1ec4a6923780443e64fff60cd8adbeaca4f1d80e0f50ecdd1a86fb6a266f6eac665ba26afa828bb5badd9844262740d78bd453057a075e5beb72a437f6a6937c941cc7bb9318c52963d0c8fd7eac17088fbafa7a0de5c5703582dadd2df0ad059e859b5c38865edbe24f57ad896eb0000000000000028c931b6d85bc74c955eba7b2da84c3972aaf8131d412dcdef7b127ec7a867d45336d27907b4408369c38f1552a6ec3e840c7bd599224f0278cb63f5f02421dd9fb9ce203f9d818bd164bd3ef114ea1c80072bae0c8d2809c9ead4b4fe6744940a1b15241bd165f974c729fc16599dd5013bbfef692d01a8ff8a1f0b7ac8f03efea80de71961d60e9c2325adeaddcbe28686de4a78b5a713a39e2f786a9e6f638ec8c8d1dbc6ca8f599a114fbc1201e1810000000000000020f763c2828f215f20407616362011799e83c511791fb78a18db13c7f785b83e15659c158be1e6837ce88162954f1c9c892bb869ec22db0154809cc9e2c9ddeada3c47928783051f20f6b56f12127fdb6900000000";
```
[the remark this answers is not in the document]
Comment (bochaco) [reply src/dbc.rs:341]: correct

File: src/dbc.rs (hunk 1 above)
Comment (bochaco) [reply src/dbc.rs:366]: it seems so, but with the change to using a set for the TXs then we should be fine

**Review by maqi — APPROVED** [review]

**Pull request merged**